File: csharp/AddOns/Indicators/EMA.cs
using System;
using System.Collections.Generic;

namespace AlgoTrading.Indicators
{
    public static class EMA
    {
        public static double Seed(IReadOnlyList<double> values, int period)
        {
            if (period <= 0) throw new ArgumentException("period must be positive, got " + period);
            if (values.Count < period)
                throw new ArgumentException("need at least " + period + " values to seed EMA, got " + values.Count);
            double sum = 0;
            for (int i = 0; i < period; i++) sum += values[i];
            return sum / period;
        }

        public static List<double?> Compute(IReadOnlyList<double> values, int period)
        {
            if (period <= 0) throw new ArgumentException("period must be positive, got " + period);
            var n = values.Count;
            var outList = new List<double?>(n);
            for (int i = 0; i < n; i++) outList.Add(null);
            if (n < period) return outList;

            double alpha = 2.0 / (period + 1);
            double seed = Seed(values, period);
            outList[period - 1] = seed;
            double prev = seed;
            for (int i = period; i < n; i++)
            {
                prev = alpha * values[i] + (1 - alpha) * prev;
                outList[i] = prev;
            }
            return outList;
        }
    }
}

File: src/algotrading/indicators/ema.py
"""Standard EMA per Appendix B §B.5.

`alpha = 2 / (N + 1)`, seeded with the simple average of the first N
closed bars.
"""

from __future__ import annotations

from collections.abc import Sequence


def ema_seed(values: Sequence[float], period: int) -> float:
    if period <= 0:
        raise ValueError(f"period must be positive, got {period}")
    if len(values) < period:
        raise ValueError(f"need at least {period} values to seed EMA, got {len(values)}")
    return sum(values[:period]) / period


def ema(values: Sequence[float], period: int) -> list[float | None]:
    """Return a list aligned to ``values``: positions < period-1 are None,
    position period-1 is the seed (simple average of first N), and from
    position period onward use the standard EMA recurrence.
    """
    if period <= 0:
        raise ValueError(f"period must be positive, got {period}")
    out: list[float | None] = [None] * len(values)
    if len(values) < period:
        return out
    alpha = 2.0 / (period + 1)
    seed = ema_seed(values, period)
    out[period - 1] = seed
    prev = seed
    for i in range(period, len(values)):
        prev = alpha * values[i] + (1 - alpha) * prev
        out[i] = prev
    return out
